File: src/point.ts
import { Direction } from "./direction";

export class Point {
  constructor(public readonly x: number, public readonly y: number) {}

  public calculate(point: Point): { length: number; direction: Direction } {
    const x1 = this.x;
    const x2 = point.x;
    const diffX = x2 - x1;
    const lengthX = Math.abs(diffX);

    const y1 = this.y;
    const y2 = point.y;
    const diffY = y2 - y1;
    const lengthY = Math.abs(diffY);

    let direction: Direction;
    let length: number;
    if (lengthX > lengthY) {
      direction = diffX > 0 ? Direction.RIGHT : Direction.LEFT;
      length = lengthX;
    } else if (lengthY >= lengthX && lengthY > 0) {
      direction = diffY > 0 ? Direction.DOWN : Direction.UP;
      length = lengthY;
    } else {
      direction = Direction.NONE;
      length = 0;
    }

    return {
      length: length,
      direction: direction,
    };
  }
}

export class PointFactory {
  public create(x: number, y: number): Point {
    return new Point(x, y);
  }

  public createForInitialize(): Point {
    return new Point(-1, -1);
  }
}

File: lua/gesture/point.lua
local M = {}

M.x_length_threshold = 5
M.y_length_threshold = 5

local Line = {}
Line.__index = Line

function Line.new(direction, length)
  local tbl = {direction = direction, length = length}
  return setmetatable(tbl, Line)
end

function Line.is_short(self)
  local length_threshold
  if self.direction == "UP" or self.direction == "DOWN" then
    length_threshold = M.y_length_threshold
  else
    length_threshold = M.x_length_threshold
  end
  return self.length < length_threshold
end

local Point = {}
Point.__index = Point

function Point.line_to(self, point)
  local x1 = self.x
  local x2 = point.x
  local diff_x = x2 - x1
  local length_x = math.abs(diff_x)

  local y1 = self.y
  local y2 = point.y
  local diff_y = y2 - y1
  local length_y = math.abs(diff_y)

  local direction, length
  if length_x > length_y then
    direction = diff_x > 0 and "RIGHT" or "LEFT"
    length = length_x
  elseif length_y >= length_x and length_y > 0 then
    direction = diff_y > 0 and "DOWN" or "UP"
    length = length_y
  else
    return nil
  end

  return Line.new(direction, length)
end

M.new = function(x, y)
  local tbl = {x = x, y = y}
  return setmetatable(tbl, Point)
end

return M
